lib.js

```diff
@@ -48,47 +47,51 @@ module.exports = (options) => {
             body: {
                 file_path: `public/.well-known/acme-challenge/${key}`,
                 commit_message: 'Automated Let\'s Encrypt renewal',
-                branch_name: 'master',
+                branch_name: repo.default_branch,
                 content: value
             }
         })).return([`http://${domain}/.well-known/acme-challenge/${key}`, value]);
     };
 
     const deleteChallenges = (key, repo) => {
         return Promise.resolve(gitlabRequest.delete({
-            url: `projects/${repo.id}/repository/files`,
+            url: `/projects/${repo.id}/repository/files`,
             body: {
                 file_path: `public/.well-known/acme-challenge/${key}`,
                 commit_message: 'Automated Let\'s Encrypt renewal',
-                branch_name: 'master'
+                branch_name: repo.default_branch
             }
         }));
     };
+
+    let deleteChallengesPromise = null;
+
     return Promise.join(getUrls, generateRsa(), generateRsa(), getRepository(options.repository),
         (urls, accountKp, domainKp, repo) => {
-            return LeCore.registerNewAccountAsync({
+            return ACME.registerNewAccountAsync({
                 newRegUrl: urls.newReg,
                 email: options.email,
-                accountPrivateKeyPem: accountKp.privateKeyPem,
+                accountKeypair: accountKp,
                 agreeToTerms: (tosUrl, cb) => {
                     console.log(`By using Let's Encrypt, you are agreeing to the TOS at ${tosUrl}`);
                     cb(null, true);
                 }
             }).then(() => {
-                return LeCore.getCertificateAsync({
+                return ACME.getCertificateAsync({
                     newAuthzUrl: urls.newAuthz,
                     newCertUrl: urls.newCert,
-                    domainPrivateKeyPem: domainKp.privateKeyPem,
-                    accountPrivateKeyPem: accountKp.privateKeyPem,
-                    domains: [options.domain],
+                    domainKeypair: domainKp,
+                    accountKeypair: accountKp,
+                    domains: options.domain,
                     setChallenge: (hostname, key, value, cb) => {
-                        return uploadChallenge(key, value, repo, options.domain)
+                        return Promise.resolve(deleteChallengesPromise)
+                            .then(() => uploadChallenge(key, value, repo, hostname))
                             .tap(res => console.log(`Uploaded challenge file, waiting for it to be available at ${res[0]}`))
                             .spread(pollUntilDeployed)
                             .asCallback(cb);
                     },
                     removeChallenge: (hostname, key, cb) => {
-                        return deleteChallenges(key, repo).asCallback(cb);
+                        return (deleteChallengesPromise = deleteChallenges(key, repo)).asCallback(cb);
```

Comment:
Because when `removeChallenge` is called, `le-acme-core` does not wait until this operation gets completed: It immediately calls `setChallenge` for the next domain instead. So I've assigned the promise that is returned from `deleteChallenges()` to variable `deleteChallengesPromise` (which is initially `null`) and then waited until it gets resolved in `setChallenge()`.